runtime/compilers/rillv1/template.go

```diff
@@ -278,3 +287,56 @@ func EvaluateBoolExpression(expr string) (bool, error) {
 	}
 	return result, nil
 }
+
+func ExtractVariablesFromTemplate(tree *parse.Tree) (map[string][]string, error) {
+	var variables []string
+	walkNodes(tree.Root, func(n parse.Node) {
+		if vn, ok := n.(*parse.FieldNode); ok {
+			variables = append(variables, joinIdentifiers(vn.Ident))
+		}
+	})
+
+	variablesMap := make(map[string][]string)
+	// Iterate through the variables and split them into key and value and add them to the map
+	for _, variable := range variables {
+		// Split the variable into parts using dot (.) as the separator
+		parts := strings.Split(variable, ".")
+		key := parts[0]
+		value := parts[1]
+
+		variablesMap[key] = append(variablesMap[key], value)
+	}
```

Comment:
Instead of building a one-level map (anyway doesn't work for deeper nesting), just return the full variable name minus the initial `.`.

So something like:
```go
if strings.HasPrefix(var, ".") {
  var = var[1:]
  variablesMap[var] = true
}

return maps.Keys(variablesMap), nil
```

Reply:
Also, this can be done directly during the walk, no need for allocating the slice `variables`

Reply:
j`oinIdentifiers(vn.Ident)` will return variables after removing the first `.`
so just need to append that and return right?

Then I guess this should work

`func ExtractVariablesFromTemplate(tree *parse.Tree) ([]string, error) {
	variablesMap := make(map[string]bool)
	walkNodes(tree.Root, func(n parse.Node) {
		if vn, ok := n.(*parse.FieldNode); ok {
			v := joinIdentifiers(vn.Ident)
			variablesMap[v] = true
		}
	})

	return maps.Keys(variablesMap), nil
}`

For template: `SELECT * FROM {{.env.partner_table_name}} WITH SAMPLING {{.env.partner_table_name}} .... {{.user.domain}}` we output should be `[]string{"env.partner_table_name", "user.domain"}`